### ISSUE TITLE
[Bash] command substitution highlighting

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#153]
Created for https://sourceforge.net/p/scintilla/feature-requests/1033/.

There is one line change to fix it (highlight variable/expansion inside string/heredoc would need to be separator issues), https://github.com/zufuliu/notepad2/commit/2230e8ced153445b93cc9eaab930347e8ed3559d
```diff
diff --git a/lexers/LexBash.cxx b/lexers/LexBash.cxx
index a071e624..24600303 100644
--- a/lexers/LexBash.cxx
+++ b/lexers/LexBash.cxx
@@ -880,7 +880,7 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
                                sc.SetState(SCE_SH_BACKTICKS);
                                QuoteStack.Start(sc.ch, BASH_DELIM_BACKTICK);
                        } else if (sc.ch == '$') {
-                               if (sc.Match("$((")) {
+                               if (sc.chNext == '(' || sc.chNext == '[') {
                                        sc.SetState(SCE_SH_OPERATOR);   // handle '((' later
                                        continue;
                                }
```

```bash
#!/bin/bash

echo $((1 + 2))
echo $[1 + 2]
echo $(date)

echo "some text" | while read -r line
do
	echo "$line"
done

output=$(echo "some text" | while read -r line
do
	echo "$line"
done)

output="$(echo "some text" | while read -r line
do
	echo "$line"
done)"
```

![image](https://user-images.githubusercontent.com/2289926/233835442-bee72391-4f30-4671-bb4b-1ede159f2aaa.png)

**Comment by nyamatongwe** [comment]
Why special case `$[` which is not mentioned in the [request ](https://sourceforge.net/p/scintilla/feature-requests/1033/)? Search the [bash documentation](https://www.gnu.org/software/bash/manual/bash.html) and `$[` doesn't occur. This may, instead, be doing something like applying `$` on a `[...]` expression.

This patch makes the code handling `$(` at line 899 unreachable.

Modifying the appearance of all `$(` is a major change.

**Comment by nyamatongwe** [comment]
The current behaviour makes it easy to see the scope of a command `$(...)` but without insight to the meaning inside the command text so there are benefits to both that and styling inside the command. That likely means a choice with a setting.

A richer presentation that provided cues for both may be worthwhile - perhaps a background colour for command string but foreground and attributes inside.
![BashCommandString](https://user-images.githubusercontent.com/225818/234129636-999f43c7-9303-45ec-af14-7645bcfa1f2a.png)

**Comment by zufuliu** [comment]
I think most code is done, will need time to split the big changes. C++ like inactive block highlighting is not implemented (which requires more changes). highlight `$(` and `)` as operator or backticks (https://sourceforge.net/p/scintilla/feature-requests/982/) is easy to implement, one can find the scope with brace matching.

![image](https://user-images.githubusercontent.com/2289926/235302113-8bdc9f71-f19b-48fc-a546-2cb6704571d1.png)

**Comment by zufuliu** [comment]
[bash-command.patch](https://github.com/ScintillaOrg/lexilla/files/11523423/bash-command.patch)

Add property to highlight inside command substitution. `$()` highlighted as `SCE_SH_OPERATOR` (similar to `$(( ))` arithmetic expansion), this is the simple than highlight it as `SCE_SH_BACKTICKS`. third option can be added to do the richer presentation highlighting.

**Comment by nyamatongwe** [comment]
Committed with 119ad899ebe33c8c0acca8a4a3e7e0ce13bdb280 but incorrectly marked as #154.

**Comment by zufuliu** [comment]
[bash-command-rich.patch](https://github.com/ScintillaOrg/lexilla/files/11530133/bash-command-rich.patch)

Draft code to implemented richer presentation highlighting (based on LexCPP), not sure how to test this but relevant `sc.SetState()`, `sc.ChangeState()` and `sc.state` are updated.

**Comment by zufuliu** [comment]
[bash-command-rich2.patch](https://github.com/ScintillaOrg/lexilla/files/11542476/bash-command-rich2.patch)

Changed to 0x40, `(streamType * numberItemTypes + itemType)` would make the code more complex and non extensible in the further when more styles are needed.

**Comment by nyamatongwe** [comment]
The term 'active' (`MaskActive`, `activitySet`) makes sense for C++ where it is saying if a range of code is active or inactive because of preprocessor state.

For bash, the state being maintained is something like `insideCommandSubstitution` so should use a name with more meaning. Could be shorter like `insideCommand` or `inCommand`.

**Comment by zufuliu** [comment]
renamed to `insideCommand` and `MaskCommand()`.
[bash-command-rich3.patch](https://github.com/ScintillaOrg/lexilla/files/11574160/bash-command-rich3.patch)

**Comment by nyamatongwe** [comment]
Committed.

Documented `lexer.bash.command.substitution=2` as provisional in case there is a need to change the values for extended styles.

**Comment by zufuliu** [comment]
> in case there is a need to change the values for extended styles

Personally l like 0x80 (which is unlike to broken in the future). the C++ lexer's use of 0x40 is likely to broken unless split it into different lexers, e.g. currently there are missing styles for `@annotation` or `@decorator`, JavaScript template literal and XML literal (JSX), C# interpolated string and raw string, Swift delimited string and regex.

**Comment by zufuliu** [comment]
SciTE's `perl.properties` is not updated with these added properties.

**Comment by nyamatongwe** [comment]
Included in release 5.2.5.
